doc/ni_by_example_3.md

```diff
@@ -1,243 +1,356 @@
+To review the syntax, the *name* of the variable is `_`, and within the body of the subroutine, the array associated with that name, `@_` is the array of values that are passed to the function.  To get any particular scalar value within, you tell Perl you want a scalar (`$`) from the variable with name `_`, and then indicate to Perl that of the variables named `_`, you want to reference the array, by using the postfix `[0]`.
+
+Subroutines can also be called without the preceding `&`, or created using the following syntax:
 
 ```
-dir_test/hi
-dir_test/there
+$ ni n1p'sub yo {"hi " . $_[0]} yo a'
+hi 1
 (END)
 ```
 
-`ni` has converted the folder into a stream of the names of files (and directories) inside it. You can thus access the files inside a directory using `\<`.
+Note that in these examples, a new function will be defined for every line in the input, which is highly inefficient. In the next section, we will introduce begin blocks, which allow functions to be defined once and then used over all of the lines in a Perl mapper block.
 
-`$ ni dir_test \<`
+####Default Variables
+While nice languages make you take pains to indicate default values and variables, Perl is not at all nice in this regard.
 
-yields:
+Consider the following spell:
 
 ```
-hello
-you
+$ ni n20p'/^(\d)\d*$/'
+```
+
+It looks like the Perl snippet is using a regular expression to take the first digit of a number. What's not clear is which number it's using. Looking at the output:
+
+```
+1
+2
+3
+4
+5
+6
+7
+8
+9
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+2
 (END)
 ```
 
-`ni` also works with the bash expansion operator `*`.
+Clearly this regex is operating on each line (the lines input were the integers 1 through 20). The way this works is another piece of Perl's uncompromising commitment to coding efficiency; default variables.
 
-For example, `ni dir_test/*` yields:
+In fact, the code above is operating on the most important of the Perl default variables, `$_`, which stores the value of the input line. Note that it shares the same name as the variable `@_`, and the similar-looking `$_[...]` will reference one of the scalars in `@_`, and not one of the characters in `$_`. Python-style string slicing (which uses square brackets) is not one available in vanilla Perl 5, where the `substr` method (or regexes) are more commonly used.
+
+
+##Intermediate Perl Operations
+####`p'^{...} ...'`: Begin Block
+A begin block is indicated by attaching a caret (`^`) to a block of code (encolsed in `{ }`). Outside of begin blocks, the Perl code is evaluated for every row; inside a Begin Block, the code is evaluated once and factored over the entire remaining Perl code. The example from the previous section:
 
 ```
-hello
-you
+$ ni n1p'sub yo {"hi " . $_[0]} yo a'
+hi 1
 (END)
 ```
 
-`ni` is able to go to the files directly becasue it is applies bash expansion first; bash expansion generates the file paths, which `ni` is then able to interpret.
+Would be better written as:
```

Comment:
This particular case isn't true, actually. `sub` is implicitly a `BEGIN`-level construct; it has to be, since function prototypes influence how Perl parses code that comes after them.